samples/boards/intel_adsp/core_power_toggle/src/main.c

```diff
@@ -0,0 +1,281 @@
+		if (arch_cpu_active(i)) {
+			mask |= BIT(i);
+		}
+	}
+
+	return mask;
+}
+
+#define NTHREADS   CONFIG_MP_MAX_NUM_CPUS
+#define STACK_SIZE 1024
+#define PRIORITY   7
+
+static struct k_sem sems[NTHREADS];
+static struct k_sem done[NTHREADS];
+static struct k_thread threads[NTHREADS];
+K_THREAD_STACK_ARRAY_DEFINE(tstack, NTHREADS, STACK_SIZE);
+
+static void add(volatile uint32_t *x, const uint32_t val)
+{
+	*x += val;
+}
+
+/* This NOP macro generates at least 2 instructions. Notably if only nop.n is
+ * used the compiler will remove duplicate nop.n instructions so a volatile add
+ * is done to ensure this does not occur.
+ */
+#define NOP(i, val, ...)                                                                           \
+	add(&val, i);                                                                              \
+	__asm__ volatile("nop.n")
+
+static bool cpu_active[NTHREADS];
+
+static void do_work(void)
+{
+	volatile uint32_t v = 0;
+
+	/* These LISTIFY macros will ensure some number of actual cpu operations
+	 * will occur to simulate real work being done.
+	 * Can be copied/pasted a few times as needed to generate more op codes.
+	 */
+	LISTIFY(4096, NOP, (;), v);
+	LISTIFY(4096, NOP, (;), v);
+}
+
+void coreN_loop(void *arg0, void *arg1, void *arg2)
+{
+	uint32_t cnt = 0;
+	uint32_t cpu = arch_curr_cpu()->id;
+
+	while (true) {
+		printk("core %d waiting on semaphore\n", cpu);
+		k_sem_take(&sems[cpu],
+			   K_FOREVER); /* put core X to sleep until semaphore is given */
+		printk("core %d doing work\n", cpu);
+		do_work();
+		printk("core %d done\n", cpu);
+		k_sem_give(&done[cpu]);
+	}
+}
+
+const char *names[4] = {
+	"core0_loop",
+	"core1_loop",
+	"core2_loop",
+	"core3_loop",
+};
+
+void core0_loop(void *arg0, void *arg1, void *arg2)
+{
+	uint32_t cnt = 0;
+	uint32_t cpu = arch_curr_cpu()->id;
+
+	for (int i = 1; i < NTHREADS; i++) {
+		k_thread_create(&threads[i], tstack[i], STACK_SIZE, coreN_loop, &sems[i], NULL,
+				NULL, K_HIGHEST_THREAD_PRIO, 0, K_FOREVER);
+		k_thread_cpu_pin(&threads[i], i);
+		k_thread_name_set(&threads[i], names[i]);
+		k_thread_start(&threads[i]);
+	}
+
+	while (true) {
+		k_sem_take(&sems[cpu], K_FOREVER);
+		for (int i = 1; i < NTHREADS; i++) {
+			printk("enabling core %d\n", i);
+			cpu_enable_core(i);
+			k_busy_wait(1000);
+			arch_sched_ipi();
+			k_sem_give(&sems[i]);
+		}
+		do_work();
+		for (int i = 1; i < NTHREADS; i++) {
+			k_sem_take(&done[i], K_FOREVER);
+			printk("disabling core %d\n", i);
+			cpu_disable_core(i);
+		}
+		k_sem_give(&done[cpu]);
+		printk("core %d thread %p loop %u\n", cpu, k_current_get(), cnt);
+		cnt++;
+	}
+}
+
+void timer_cb(struct k_timer *t)
+{
+	printk("core %d timer\n", arch_curr_cpu()->id);
+
+	k_sem_give(&sems[0]);
+}
+
+K_TIMER_DEFINE(timer0, timer_cb, NULL);
+
+#if defined(CONFIG_PM_POLICY_CUSTOM)
+
+#define PLATFORM_PRIMARY_CORE_ID 0
+
+const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
```

Comment:
I think this custom policy on SoF really not needed. I am pushing some changes into intel_adsp to avoid some of checks that are being done here. The missing bits could be done just putting / releasing constraints properly.

Reply:
That's great if so, I was trying to replicate this as close as possible to what SoF has